federation/executor.go

```diff
@@ -0,0 +1,323 @@
+package federation
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/samsarahq/thunder/graphql"
+	"github.com/samsarahq/thunder/graphql/introspection"
+	"golang.org/x/sync/errgroup"
+)
+
+type ExecutorClient interface {
+	Execute(ctx context.Context, req *graphql.Query) ([]byte, error)
+}
+
+// Executor has a map of all the executor clients such that it can execute a 
+// subquery on any of the federated servers.
+// The planner allows it to coordinate the subqueries being sent to the federated servers
+type Executor struct {
+	Executors map[string]ExecutorClient
+	planner   *Planner  
+}
+
+func fetchSchema(ctx context.Context, e ExecutorClient) ([]byte, error) {
+	query, err := graphql.Parse(introspection.IntrospectionQuery, map[string]interface{}{})
+	if err != nil {
+		return nil, err
+	}
+
+	return e.Execute(ctx, query)
+}
+
+
+func NewExecutor(ctx context.Context, executors map[string]ExecutorClient) (*Executor, error) {
+	// Fetches the schemas from the executors clients
+	schemas := make(map[string]*introspectionQueryResult)
+	for server, client := range executors {
+		schema, err := fetchSchema(ctx, client)
+		if err != nil {
+			return nil, fmt.Errorf("fetching schema %s: %v", server, err)
+		}
+
+		var iq introspectionQueryResult
+		if err := json.Unmarshal(schema, &iq); err != nil {
+			return nil, fmt.Errorf("unmarshaling schema %s: %v", server, err)
+		}
+
+		schemas[server] = &iq
+	}
+
+	types, err := convertSchema(schemas)
+	if err != nil {
+		return nil, fmt.Errorf("converting schema error: %v", err)
+	}
+
+	introspectionSchema := introspection.BareIntrospectionSchema(types.Schema)
+	introspectionServer := &Server{schema: introspectionSchema}
+
+	executors["introspection"] = &DirectExecutorClient{Client: introspectionServer}
```

Comment:
consider making the key string a const and maybe adding a check that the `executors` map doesn't already contain this key?